xarray/coding/variables.py

```diff
@@ -235,10 +235,19 @@ def encode(self, variable: Variable, name: T_Name = None):
                 f"Variable {name!r} has conflicting _FillValue ({fv}) and missing_value ({mv}). Cannot encode data."
             )
 
+        # cast to correct dtype in case of times
+        # see GH 7817
+        units = attrs.get("units", None)
+        if isinstance(units, str) and "since" in units:
+            encoded_dtype = encoding.pop("dtype")
+            if encoded_dtype is not None and encoded_dtype != data.dtype:
+                data = np.asarray(data, dtype=encoded_dtype)
```

Comment:
Ah, I'll see how to make this lazy. Thanks.

Reply:
@Illviljan I've checked the other encoders:

- `CFDatetimeCoder`/`CFTimedeltaCoder` using `np.asarray` https://github.com/pydata/xarray/blob/306310050f6278f4d8107159c9c3649fbd6b5dea/xarray/coding/times.py#L636
- `CFScaleOffsetCoder` uses `data.astype` https://github.com/pydata/xarray/blob/306310050f6278f4d8107159c9c3649fbd6b5dea/xarray/coding/variables.py#L333
- `UnsignedIntegerCoder` uses `duck_array_ops.astype` https://github.com/pydata/xarray/blob/306310050f6278f4d8107159c9c3649fbd6b5dea/xarray/coding/variables.py#L380
- `NonStringCoder` uses `data.astype` https://github.com/pydata/xarray/blob/306310050f6278f4d8107159c9c3649fbd6b5dea/xarray/coding/variables.py#L510
- `BooleanCoder` uses `duck_array_ops.astype` https://github.com/pydata/xarray/blob/306310050f6278f4d8107159c9c3649fbd6b5dea/xarray/coding/variables.py#L452

Now I'm a bit confused what to choose, maybe simple `.astype` is good enough? As `CFDatetimeCoder` is first run and already uses `np.asarray` this might work as is. Anyway to me it looks like those different casts might be aligned or they all serve a different purpose.

Reply:
But the function has a Variable as an input, and that can be any duckarray (sparse, dask, etc)

I think you should go with `duck_array_ops.astype`. array api standard seems to only have `xp.astype`, https://data-apis.org/array-api/latest/API_specification/generated/array_api.astype.html.

Reply:
Thanks @Illviljan, much appreciated.